tests/utils/test_msa_aligner.py

```diff
@@ -0,0 +1,145 @@
+        mafft._run_aligner(args, env)
+
+        popen_mock.assert_called_once_with(args, stderr=subprocess.PIPE, encoding="utf-8", shell=True, env=env)
+        wait_mock.wait.assert_called_once_with()
+
+    @patch("shutil.which", return_value=1)
+    @patch.object(Path, Path.mkdir.__name__)
+    @patch("subprocess.Popen")
+    def test___run_aligner___fail_run(self, popen_mock, *uninteresting_mocks):
+        wait_mock = Mock()
+        wait_mock.wait.return_value = 1
+        popen_mock.return_value = wait_mock
+        args = "fake_arg_1 fake_arg_2 fake_arg_3"
+        env = {"fake_key_1": "fake_value_1", "fake_key_2": "fake_value_2"}
+
+        mafft = MAFFT(self.executable, self.tmpdir)
+        with self.assertRaises(ExecutionError):
+            mafft._run_aligner(args, env)
+
+        popen_mock.assert_called_once_with(args, stderr=subprocess.PIPE, encoding="utf-8", shell=True, env=env)
+        wait_mock.wait.assert_called_once_with()
+
+    @patch("shutil.which", return_value=1)
+    @patch.object(Path, Path.mkdir.__name__)
+    def test___create_new_sequences_file(self, *uninteresting_mocks):
+        directory = Path("tests/data/utils/create_new_sequences_file")
+        new_sequences = {"ACGT", "AAAA", "TTTT"}
+
+        mafft = MAFFT(self.executable, self.tmpdir)
+        mafft._create_new_sequences_file(directory, new_sequences)
+
+        self.assertTrue(filecmp.cmp(directory/"new_sequences.fa", directory/"new_sequences.truth.fa"))
+
+    @patch("shutil.which", return_value=1)
+    @patch.object(Path, Path.mkdir.__name__)
+    def test___get_aligner_name(self, *uninteresting_mocks):
+        mafft = MAFFT(self.executable, self.tmpdir)
+
+        expected = "MAFFT"
+        actual = mafft.get_aligner_name()
+
+        self.assertEqual(expected, actual)
+
+    @patch("shutil.which", return_value=1)
+    @patch.object(Path, Path.mkdir.__name__)
+    @patch("shutil.rmtree")
+    def test___cleanup_run(self, shutil_rmtree_mock, *uninteresting_mocks):
+        run_tmpdir = Path("run_tmpdir")
+
+        mafft = MAFFT(self.executable, self.tmpdir)
+        mafft._cleanup_run(run_tmpdir)
+
+        shutil_rmtree_mock.assert_called_once_with(run_tmpdir)
+
+    @patch("shutil.which", return_value=1)
+    @patch.object(MAFFT, MAFFT._create_new_sequences_file.__name__, return_value=Path("_create_new_sequences_file"))
+    @patch("os.environ", {"fake_key_1": "fake_value_1"})
+    @patch.object(Path, Path.mkdir.__name__)
+    @patch("make_prg.utils.msa_aligner.create_temp_dir", return_value=Path("fake_tmp_dir"))
+    @patch("builtins.open", new_callable=mock_open)
+    @patch.object(SeqIO, SeqIO.write.__name__)
+    @patch.object(MAFFT, MAFFT._run_aligner.__name__)
+    @patch("make_prg.utils.msa_aligner.load_alignment_file", return_value="updated_alignment")
+    @patch.object(MAFFT, MAFFT._cleanup_run.__name__)
+    def test___get_updated_alignment(self, cleanup_run_mock, load_alignment_file_mock, run_aligner_mock, write_mock,
```

Comment:
this test is very complicated, and is essentially only testing that the function executes, without checking actual mafft outputs right?

i'm not sure i see this test's value

Reply:
sorry don't mean to be brutal it looks like it was hard to engineer

Reply:
No worries, you were not brutal :p

It basically tests the whole [get_updated_alignment() function](https://github.com/leoisl/make_prg/blob/2e2ae14f74105e3d10f104ab53785fd28fba9309/make_prg/utils/msa_aligner.py#L88-L124), but by mocking heavy function calls, like creating temp dir for MAFFT to run in, creating MAFFT input files, running MAFFT itself, loading the updated alignment and cleaning up the run. This is indeed a complex function to test in the sense that it does lots of heavy stuff. The approach I followed here is mocking all these heavy stuff, as well as calls to other functions that were already unit tested, as I don't think we should require the user to have MAFFT installed in order to test this function, we can just mock it, that is what is usually done when you have a heavy dependency (e.g. a DB connection) during a test. But just mocking is not enough, as if we just mock the MAFFT call, but we invoke it with wrong parameters, the tests still would pass, whereas they should not. We need to make sure that the mocked calls received the parameters the way we expect, and that we know would work in a real execution, and this is asserted [by the end of the test case](https://github.com/leoisl/make_prg/blob/2e2ae14f74105e3d10f104ab53785fd28fba9309/tests/utils/test_msa_aligner.py#L132-L145)

About the value of this test case, IDK actually how to evaluate it, but I think it is worth keeping it...

Reply:
yep works!